### ISSUE TITLE
DRY Condition checking

### ISSUE THREAD

**Issue by rainest** [Kong/kubernetes-ingress-controller#2609]
Gateway resources use [Conditions](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#Condition) extensively to indicate various status states. Various parts of the controller code need to check Conditions to see whether they can proceed or whether test criteria are satisfied.

Conditions are stored in an arbitrarily-ordered slice, so in order to perform these checks, we have a number of similar loops throughout the code (`git grep -nr Conditions | grep range`) to find the particular Condition type we want and check that its status and reason are the expected value, e.g.:

```
func isGatewayScheduled(gateway *gatewayv1alpha2.Gateway) bool {
	for _, cond := range gateway.Status.Conditions {
		if cond.Type == string(gatewayv1alpha2.GatewayConditionScheduled) &&
			cond.Reason == string(gatewayv1alpha2.GatewayReasonScheduled) &&
			cond.Status == metav1.ConditionTrue {
			return true
		}
	}
	return false
}
```

```
for _, lstatus := range gw.Status.Listeners {         
    ready := false        
    for _, condition := range lstatus.Conditions {        
        if condition.Type == string(gatewayv1alpha2.ListenerConditionReady) && condition.Status == metav1.ConditionTrue {        
            ready = true        
        }        
    }        
    if !ready {        
        return false        
    }        
}        
return true
```

These should be possible to replace with a generic function, unit testable function that takes a Condition slice and desired Type, Reason, and Status values as input and returns whether such a Condition is present or not.

## Acceptance Criteria

- [ ] Code repetition mitigated for the case described in the issue.

**Comment by pmalek** [comment]
> May be resolved by #2609

This is #2609 😅 Recursion overload

**Comment by scseanchow** [comment]
> > May be resolved by #2609
> 
> This is #2609 😅 Recursion overload

haha, fixed: https://github.com/Kong/kubernetes-ingress-controller/issues/2694

**Comment by czeslavo** [comment]
@randmonkey regarding the discussion above - as you resolved #2694 with https://github.com/Kong/kubernetes-ingress-controller/pull/2791 - is this ticket still valid in your opinion?

**Comment by randmonkey** [comment]
I think this issue is still valid because we still have functions like `isGatewayScheduled` and `isGatewayReady` to check the conditions of `Gateway`s. The purpose of this issue is to use a unified way to check whether the condition of `Gateway` is in a desired state. #2694 was a prerequisite of this.
